### ISSUE TITLE
[CUDAExecutionProvider] Regression from ORT 1.15.0 onwards: Compute MatMul dimension mismatch

### ISSUE THREAD

**Issue by fxmarty** [microsoft/onnxruntime#18692]
### Describe the issue

Hi, I noticed a regression in `onnxruntime-gpu==1.15.1` and  `onnxruntime-gpu==1.16.3` (no problem on `onnxruntime-gpu==1.14.1`.

The following code runs fine on `CPUExecutionProvider` for all three ORT versions, but fails on `CUDAExecutionProvider` for `1.15.1` and `1.16.3`.

```python
import onnxruntime
from transformers import DetrImageProcessor
import torch
from PIL import Image
import requests
from transformers.models.detr.modeling_detr import DetrObjectDetectionOutput

from optimum.onnxruntime import ORTModelForCustomTasks

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

model = onnxruntime.InferenceSession("/path/to/model.onnx", providers=["CUDAExecutionProvider"])

processor = DetrImageProcessor.from_pretrained("facebook/detr-resnet-50", revision="no_timm")

inputs = processor(images=image, return_tensors="np")
inputs = {
    "pixel_values": inputs["pixel_values"]
}

outputs = model.run(None, inputs)
```

with the error:
```
Traceback (most recent call last):
  File "<tmp 1>", line 22, in <module>
    outputs = model.run(None, inputs)
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 217, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running MatMul node. Name:'/model/decoder/layers.0/self_attn/out_proj/MatMul' Status Message: matmul_helper.h:59 Compute MatMul dimension mismatch
```

### To reproduce

As above. Reproduce with https://huggingface.co/fxmarty/bugged-detr-ort-cuda/tree/main

Using CUDA 11.7, which should be compatible according to https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html

### Urgency

medium

### Platform

Linux

### OS Version

Ubuntu 22.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

as above

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.7

**Comment by fxmarty** [comment]
cc @tianleiwu @yufenglee

**Comment by fxmarty** [comment]
Hi @yufenglee @tianleiwu, this issue is not stale and reported by user for an other architecture (table-transformer): https://github.com/huggingface/optimum/issues/1774

**Comment by tianleiwu** [comment]
The issue is resolved in the main branch.

I did reproduce it in 1.17.1:
* The issue is gone after disabling all graph optimizations in session options.
* I changed graph optimization level to basic, the issue is still there.
So the issue is caused by some basic level graph optimization. 
If there is time, some debugging (by disabling basic level graph optimization one by one) can find which optimizer is the cause.
